properties/pkg/utils.go

```diff
@@ -0,0 +1,287 @@
+		// Set previous index by making sure that it's value is inbound
+		if idx > 0 {
+			prevIdx = idx - 1
+		}
+		// Iterate over the separator characters.
+		for _, sp := range sep {
+			// If there is a separator match and the previous is an escape character
+			// then copy data up to the previous index to the new string then append
+			// the current leaving out the previous character.
+			if c == sp && orig[prevIdx] == EscapeChar {
+				newSlice = append(newSlice, orig[startIdx:prevIdx]...)
+				newSlice = append(newSlice, c)
+				startIdx = idx + 1
+				break
+			}
+		}
+	}
+	// Make sure that all the original string is copied to the new
+	if startIdx <= (length - 1) {
+		newSlice = append(newSlice, orig[startIdx:]...)
+	}
+
+	return string(newSlice)
+}
+
+// EscapeSeparators returns the given s string having the Separators escaped.
+func EscapeSeparators(s string) string {
+	// Get the length of the s string
+	length := len(s)
+
+	// Return immediately if s string is empty
+	if length == 0 {
+		return s
+	}
+
+	// Convert s string to slice of rune
+	orig := []rune(s)
+	// Convert Separators to slice of rune
+	sep := []rune(Separators)
+	// Create new slice holding the escaped string
+	newSlice := make([]rune, 0)
+
+	// Index from where we need to copy the data to the new slice
+	startIdx := 0
+	// Track previous index
+	prevIdx := 0
+
+	// Iterate over the original string to find separator characters
+	for idx, c := range orig {
+		// Set previous index by making sure that it's value is inbound
+		if idx > 0 {
+			prevIdx = idx - 1
+		}
+		// Iterate over the separator characters
+		for _, sp := range sep {
+			// If there is a separator match and the previous is not an escape character
+			// then copy data up to the current index to the new string then append the
+			// escape and the current characters.
+			if c == sp && orig[prevIdx] != EscapeChar {
+				newSlice = append(newSlice, orig[startIdx:idx]...)
+				newSlice = append(newSlice, []rune{EscapeChar, c}...)
+				startIdx = idx + 1
+				break
+			}
+		}
+	}
+	// Make sure that all the original string is copied to the new
+	if startIdx <= (length - 1) {
+		newSlice = append(newSlice, orig[startIdx:]...)
+	}
+
+	return string(newSlice)
+}
+
+// GetSeparator return the separator character and its index if it is fond in the given string.
+// Otherwise an error is returned.
+func GetSeparator(s string) (string, int, error) {
+	// Index of the detected separator.
+	var sepIdx int
+	// Detected separator.
+	var sep string
+
+	// Get the length of the s string
+	length := len(s)
+
+	// Return immediately if s string is empty
+	if length == 0 {
+		return sep, sepIdx, errors.New("properties: no separator found in empty string")
+	}
+
+	// Convert s string to slice of rune
+	r := []rune(s)
+	// Convert Separators to slice of rune
+	separators := []rune(Separators)
+
+	// Track previous index
+	prevIdx := 0
+
+	// Iterate over the input string
+	for idx, c := range r {
```

Comment:
I gave a thought to your suggestions about using `s.IndexRune()` and I came to the conclusion that it would make the logic less efficient compared to how it is implemented currently.
Using `s.IndexRune()` would require to iterate over the string 3 times as there are 3 valid separator characters to find and we need to pick the one with the smallest index. Additionally we need to make sure that found separator is no escaped by checking the previous char.
The current solution is aligned with these requirements, therefore I prefer to keep it as is for now even if I do agree that copying the original string to `[]rune` is probably avoidable.
I will take another look on this and will create a separate PR if I find a better solution.
Anyways, thanks for bringing up this! 🙏